### DISCUSSION TITLE
Face some problems on receiving sound.

### DISCUSSION THREAD

**Discussion by ech-o-o** [sist-cs120/project-wiki#29] [Project 2 Q&A]
C++, with the use of ASIO4ALLv2.15 + JUCE

I met some problems so want to remove carrier sound and transmit the 1/0s straight forward, so I test like this:

For speaker: 
void audioDeviceIOCallback(const float** inputChannelData, int numInputChannels,
        float** outputChannelData, int numOutputChannels, int numSamples) override {
        int count_o = 0;
        int count_n = 0;
        for (int i = 0; i < numSamples; ++i)
        {
           if (i < 100)
            {
                count_o += 1;
                outputChannelData[0][1] = 1;
            }
            else
            {
                outputChannelData[0][i] = -1;
                count_n += 1;
            }
        std::cout << count_n << " " << count_o << std::endl;
    }

For mics:
void audioDeviceIOCallback(const float** inputChannelData, int numInputChannels,
        float** outputChannelData, int numOutputChannels, int numSamples) override {
        int count_o = 0;
        int count_n = 0;
        for (int i = 0; i < numSamples; ++i)
        {
            if (inputChannelData[0][i] < 0)
                count_n += 1;
            else
                count_o += 1;
        std::cout << count_n << " " << count_o << std::endl;
    }

From my point of view, even if time shift, they should get same number of 1 and -1, but the result is, for mics,  it outputs 412, 100. For speaker, it outputs stabled 238 274. Is my understanding wrong or some settings wrong with my computer?

Another problem is, if I set all outputs to 1, the receiver will receive nothing. I guess it ignore direct current... Or there is a solution?

P.S, the cable is correctly connected and test is not running under charging. From ASIO4ALL, it tells both computer use 48kHz sample rate, 512 buffer,  and I set all the other thing same, equal to zero.

**Comment by ech-o-o** [comment]
The code has a little bug. For speaker, outputChannelData[0][1] should be outputChannelData[0][i], I 'm sorry that I didn't check my post, but the result ran in the correct version.

**Comment by yangfengxu12** [comment]
It is the result of the mic and speaker look like reversed. The speaker should be 100, 412 and the mic should be 238 274. It is right?

**Reply by ech-o-o** [reply]
I solve this problem by extend the buffer, and this make output result stable enough for decoding...
I am not sure if you mean it is when I do "cout", I reverse the count so my output does not match... The problem I thought was that the latency should not affect the number of positives in one time frame, It just affect when a data is received. If latency will change the value, there should have no way to synchronize the signals. I do not know why this occurs but if I enlarge the buffer and signal sample rate, which means duplicate a signal more, it can be solved. I am not sure if this will help somebody but for those whose buffer size is default to be 512, you can enlarge it to get better result.
